README.md
# Ruby Sierra API Client

Simple client for querying the Sierra API

## Version

> 1.3.0

## Using

1. Configure the client with Sierra API credentials
2. Make requests

### Configuration

Example configuration:

```ruby
require 'nypl_sierra_api_client'

client = SierraApiClient.new({
  base_url: "https://[fqdn]/iii/sierra-api/v5", # Defaults to ENV['SIERRA_API_BASE_URL']
  client_id: "client-id", # Defaults to ENV['SIERRA_OAUTH_ID']
  client_secret: "client-secret", # Defaults to ENV['SIERRA_OAUTH_SECRET']
  oauth_url: "https://[fqdn]/iii/sierra-api/v3/token", # Defaults to ENV['SIERRA_OAUTH_URL'],
  log_level: "debug" # Defaults to 'info'
})
```

### Requests

Example GET:

```ruby
bib_response = sierra_client.get 'bibs/12345678'
bib = bib_response.body
```

Example POST:

```ruby
check_login = sierra_client.post 'patrons/validate', { "barcode": "1234", "pin": "6789" }
valid = check_login.success?
invalid = check_login.error?
```

Note that only GET and POST are supported at writing.

### Responses

Because of the variety of HTTP status codes and "Content-Type"s returned by the Sierra REST API, the Sierra API Client makes few assumptions about the response. All calls return a `SierraApiResponse` object with the following methods:

 * `code`: HTTP status code as an Integer (e.g. `200`, `500`)
 * `success?`: True if `code` is 2** or 3**
 * `error?`: True if `code` is >= `400`
 * `body`: The returned body. If response header indicates it's JSON, it will be a `Hash`.
 * `response`: The complete [`Net::HTTPResponse`](https://ruby-doc.org/stdlib-2.7.1/libdoc/net/http/rdoc/Net/HTTPResponse.html) object for inspecting anything else you like.

In the spirit of agnostism, the client will not intentionally raise an error when it encounters an error HTTP status code. Client will only raise an error when the request could not be carried out as specified and should be retried, that is:
 - Network failure
 - Invalid token error (401) once maximum (3) retry attempts met

## Contributing

This repo uses a single, versioned `master` branch.

 * Create feature branch off `master`
 * Compute next logical version and update `README.md`, `CHANGELOG.md`, & `nypl_sierra_api_client.gemspec`
 * Create PR against `master`
 * After merging the PR, git tag `master` with new version number.

### Updating gem

After merging to `master`, push the updated gem to rubygems.org:

```
gem build nypl_sierra_api_client.gemspec
gem push nypl_sierra_api_client-[version].gem
```

See [this guide](https://guides.rubygems.org/make-your-own-gem/) for additional help.

## Testing

```
bundle exec rspec
```


CHANGELOG.md
# Changelog

## v1.3.0

Add reattempted requests with exponential backoff, logging, and erroring for empty responses from Sierra
Add exponential backoff for authentication reattempts

## v1.2.0

Add PUT and DELETE functionality 

## v1.1.0

Add in OAuth re-authentication, up to three successive attempts, for 401 response from Sierra API for an authenticated request.

## v1.0.3

Fix: Critical bugs in initializer

## v1.0.2

Fix: Correct use of Net:HTTP constructor, where previously we were dropping the query string

## v1.0.1

Fix: Rename core library to match gem name for consistent `require`

## v1.0.0

Initial push with support for:
 - Arbitrary GET and POST requests


spec/spec_helper.rb
require 'json'
require 'nypl_log_formatter'

require_relative '../lib/nypl_sierra_api_client'

ENV['LOG_LEVEL'] ||= 'error'
ENV['APP_ENV'] = 'test'

def load_fixture (file)
  JSON.parse File.read("./spec/fixtures/#{file}")
end


spec/sierra_api_client_spec.rb
require 'spec_helper'
require 'webmock/rspec'

describe SierraApiClient do
  before(:each) do
    ENV['SIERRA_API_BASE_URL'] = 'https://example.com/iii/'
    ENV['SIERRA_OAUTH_ID'] = Base64.strict_encode64 'fake-client'
    ENV['SIERRA_OAUTH_SECRET'] = Base64.strict_encode64 'fake-secret'
    ENV['SIERRA_OAUTH_URL'] = 'https://example.com/token'

    stub_request(:post, "#{ENV['SIERRA_OAUTH_URL']}")
      .to_return(status: 200, body: '{ "access_token": "fake-access-token" }')
    stub_request(:get, "#{ENV['SIERRA_API_BASE_URL']}patrons/12345")
      .to_return({
        status: 200,
        body: '{ "id": "12345" }',
        headers: { 'Content-Type' => 'application/json;charset=UTF-8' }
      })
  end

  describe :config do
    it "should throw error if api base url unset" do
      ENV['SIERRA_API_BASE_URL'] = nil
      expect { SierraApiClient.new }.to raise_error(SierraApiClientError)
    end

    it "should throw error if client id unset" do
      ENV['SIERRA_OAUTH_ID'] = nil
      expect { SierraApiClient.new }.to raise_error(SierraApiClientError)
    end

    it "should throw error if client secret unset" do
      ENV['SIERRA_OAUTH_SECRET'] = nil
      expect { SierraApiClient.new }.to raise_error(SierraApiClientError)
    end

    it "should throw error if oauth url unset" do
      ENV['SIERRA_OAUTH_URL'] = nil
      expect { SierraApiClient.new }.to raise_error(SierraApiClientError)
    end

    it "should allow configuration via constructor" do
      ENV['SIERRA_API_BASE_URL'] = nil
      ENV['SIERRA_OAUTH_ID'] = nil
      ENV['SIERRA_OAUTH_SECRET'] = nil
      ENV['SIERRA_OAUTH_URL'] = nil

      config = SierraApiClient.new({
        base_url: 'https://example.com/iii/',
        client_id: 'client-id',
        client_secret: 'client-secret',
        oauth_url: 'https://example.com/token'
      }).instance_variable_get(:@config)

      expect(config).to be_a(Hash)
      expect(config[:base_url]).to eq('https://example.com/iii/')
      expect(config[:client_id]).to eq('client-id')
    end

    it "should prefer constructor config over ENV variables" do
      client = SierraApiClient.new({
        client_id: 'client-id-via-constructor-config'
      })
      expect(client.instance_variable_get(:@config)).to be_a(Hash)
      expect(client.instance_variable_get(:@config)[:client_id]).to eq('client-id-via-constructor-config')
    end

    it "should set default log_level to 'info'" do
      client = SierraApiClient.new
      expect(client.instance_variable_get(:@config)).to be_a(Hash)
      expect(client.instance_variable_get(:@config)[:log_level]).to eq('info')
    end

    it "should allow log_level override via constructor (only)" do
      ENV['LOG_LEVEL'] = 'debug'
      client = SierraApiClient.new
      expect(client.instance_variable_get(:@config)).to be_a(Hash)
      expect(client.instance_variable_get(:@config)[:log_level]).to eq('info')

      client = SierraApiClient.new(log_level: 'debug')
      expect(client.instance_variable_get(:@config)).to be_a(Hash)
      expect(client.instance_variable_get(:@config)[:log_level]).to eq('debug')
    end
  end

  describe :parse_http_options do
    it "should assume common defaults" do
      options = SierraApiClient.new.send :parse_http_options, {}

      expect(options[:authenticated]).to eq(true)
      expect(options[:headers]).to be_a(Hash)
      expect(options[:headers].keys.size).to eq(0)
    end

    it "should allow authentication override" do
      options = SierraApiClient.new.send :parse_http_options, { authenticated: false }

      expect(options[:authenticated]).to eq(false)
    end

    it "should allow custom Content-Type" do
      options = SierraApiClient.new.send :parse_http_options, { headers: { 'Content-Type' => 'text/plain' } }

      expect(options[:headers]).to be_a(Hash)
      expect(options[:headers]['Content-Type']).to be_a(String)
      expect(options[:headers]['Content-Type']).to eq('text/plain')
    end

    it "should allow extra header" do
      options = SierraApiClient.new.send :parse_http_options, { headers: { 'X-My-Header': 'header value' } }

      expect(options[:authenticated]).to eq(true)
      expect(options[:headers]).to be_a(Hash)
      expect(options[:headers]['X-My-Header']).to eq('header value')
    end
  end

  describe :authentication do

    it "should authenticate by default" do
      client = SierraApiClient.new

      # Verify no access token:
      expect(client.instance_variable_get(:@access_token)).to be_nil

      # Call an endpoint with authentication:
      expect(client.get('patrons/12345')).to be_a(Object)

      # Verify access_token retrieved:
      expect(client.instance_variable_get(:@access_token)).to be_a(String)
      expect(client.instance_variable_get(:@access_token)).to eq('fake-access-token')
    end

    it "should authenticate when calling with :authenticated => true" do
      client = SierraApiClient.new

      # Verify no access token:
      expect(client.instance_variable_get(:@access_token)).to be_nil

      # Call an endpoint with authentication:
      expect(client.get('patrons/12345', authenticated: true)).to be_a(Object)

      # Verify access_token retrieved:
      expect(client.instance_variable_get(:@access_token)).to be_a(String)
      expect(client.instance_variable_get(:@access_token)).to eq('fake-access-token')
    end

    it "should NOT authenticate when calling with :authenticated => false" do
      client = SierraApiClient.new

      # Verify no access token:
      expect(client.instance_variable_get(:@access_token)).to be_nil

      # Call an endpoint without authentication:
      expect(client.get('patrons/12345', authenticated: false)).to be_a(Object)

      # Verify access_token retrieved:
      expect(client.instance_variable_get(:@access_token)).to be_nil
    end
  end

  describe :do_request do
    it "should throw if invalid method" do
      expect { SierraApiClient.new.send :do_request, 'patch', 'some-path' }.to raise_error(SierraApiClientError)
    end

    it "should perform get" do
      stub_request(:get, "#{ENV['SIERRA_API_BASE_URL']}some-path")
        .to_return({
          status: 200,
          body: '{ "foo": "bar" }',
          headers: { 'Content-Type' => 'application/json;charset=UTF-8' }
        })

      resp = SierraApiClient.new.send :do_request, 'get', 'some-path'
      expect(resp).to be_a(SierraApiResponse)
      expect(resp.body).to be_a(Hash)
      expect(resp.body['foo']).to eq('bar')
    end

  end

  describe :responses do
    it "should auto parse JSON if Content-Type is json and 200 response" do
      stub_request(:get, "#{ENV['SIERRA_API_BASE_URL']}some-path")
        .to_return({
          status: 200,
          body: '{ "foo": "bar" }',
          headers: { 'Content-Type' => 'application/json;charset=UTF-8' }
        })

      resp = SierraApiClient.new.get('some-path')
      expect(resp).to be_a(SierraApiResponse)
      expect(resp.body).to be_a(Hash)
      expect(resp.body['foo']).to eq('bar')
    end

    it "should not auto parse JSON if response Content-Type is other than application/json" do
      stub_request(:get, "#{ENV['SIERRA_API_BASE_URL']}some-path")
        .to_return({
          status: 404,
          body: '{ "foo": "bar" }',
          headers: {
            'Content-Type' => 'text/html; charset=UTF-8'
          }
        })

      resp = SierraApiClient.new.get('some-path')
      expect(resp).to be_a(SierraApiResponse)
      expect(resp.body).to be_a(String)
      expect(JSON.parse(resp.body)).to be_a(Hash)
      expect(JSON.parse(resp.body)['foo']).to eq('bar')
    end

    it "should throw SierraApiClientError if response is not valid json" do
      stub_request(:get, "#{ENV['SIERRA_API_BASE_URL']}some-path")
        .to_return({
          status: 200,
          body: '{ ""mangled json }',
          headers: {
            'Content-Type' => 'application/json;charset=UTF-8'
          }
        })
      expect { SierraApiClient.new.get('some-path').body }.to raise_error(SierraApiResponseError)
    end

    it "should not throw errors for 500" do
      stub_request(:get, "#{ENV['SIERRA_API_BASE_URL']}some-path").to_return(status: 500, body: '' )

      resp = SierraApiClient.new.get('some-path')
      expect(resp).to be_a(SierraApiResponse)
      expect(resp.code).to eq(500)
    end

    it "should retry requests up to three times when given an empty response object" do 
      stub_request(:get, "#{ENV['SIERRA_API_BASE_URL']}some-path")
        .to_return({
          status: 200,
          body: '',
          headers: { 'Content-Type' => 'application/json;charset=UTF-8' }
        })

        client = SierraApiClient.new
        expect { client.get('some-path') }.to raise_error(SierraApiResponseError)
        assert_requested(
          :get,
          "#{ENV['SIERRA_API_BASE_URL']}some-path",
          times: 4
        )
        expect(client.instance_variable_get(:@retries)).to eq(0)
    end 
  end

  describe :refresh_oauth do
    before do
      stub_request(:get, "#{ENV['SIERRA_API_BASE_URL']}one-reattempt").to_return(
        { status: 401, body: '{ "foo": "bar1" }' },
        { status: 200, body: '{ "foo": "bar1" }' },
      )

      stub_request(:get, "#{ENV['SIERRA_API_BASE_URL']}maximum-attempts-path").to_return(status: 401, body: '{ "foo": "bar1" }')

      stub_request(:post, "#{ENV['SIERRA_OAUTH_URL']}")
        .to_return(
          { status: 200, body: '{ "access_token": "fake-access-token" }' },
          { status: 200, body: '{ "access_token": "second-fake-access-token" }' },
        )
    end

    it "should refresh oauth token for 401" do
      client = SierraApiClient.new
      first_token = client.instance_variable_get(:@access_token)

      resp = client.get('one-reattempt')
      second_token = client.instance_variable_get(:@access_token)

      expect(first_token).not_to eq(second_token)
      expect(client.instance_variable_get(:@retries)).to eq(0)
      expect(resp).to be_a(SierraApiResponse)
      expect(resp.code).to eq(200)
    end

    it "should throw error once maximum retries attempted" do
      client = SierraApiClient.new
      expect { client.get('maximum-attempts-path') }.to raise_error(SierraApiClientTokenError)
      assert_requested(
        :get,
        "#{ENV['SIERRA_API_BASE_URL']}maximum-attempts-path",
        times: 4
      )
      expect(client.instance_variable_get(:@retries)).to eq(0)
    end
  end
end


lib/sierra_api_client.rb
require 'net/http'
require 'net/https'
require 'uri'
require 'nypl_log_formatter'

require_relative 'errors'
require_relative 'sierra_api_response'

class SierraApiClient
  def initialize(config = {})
    config_defaults = {
      env: {
        base_url: 'SIERRA_API_BASE_URL',
        client_id: 'SIERRA_OAUTH_ID',
        client_secret: 'SIERRA_OAUTH_SECRET',
        oauth_url: 'SIERRA_OAUTH_URL'
      },
      static: {
        log_level: 'info'
      }
    }

    @config = config_defaults[:env].map {|k,v| [k, ENV[k]]}.to_h
      .merge config_defaults[:static]
      .merge config

    config_defaults[:env].each do |key, value|
      raise SierraApiClient.new "Missing config: neither config.#{key} nor ENV.#{value} are set" unless @config[key]
    end
  end

  def get (path, options = {})
    options = parse_http_options options

    do_request 'get', path, options
  end

  def post (path, body, options = {})
    options = parse_http_options options

    # Default to POSTing JSON unless explicitly stated otherwise
    options[:headers]['Content-Type'] = 'application/json' unless options[:headers]['Content-Type']

    do_request 'post', path, options do |request|
      request.body = body
      request.body = request.body.to_json unless options[:headers]['Content-Type'] != 'application/json'
    end
  end

  private

  def do_request (method, path, options = {})
    # For now, these are the methods we support:
    raise SierraApiClientError, "Unsupported method: #{method}" unless ['get', 'post'].include? method.downcase

    authenticate! if options[:authenticated]

    uri = URI.parse("#{@config[:base_url]}#{path}")

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme === 'https'

    # Build request headers:
    request_headers = {}
    request_headers['Content-Type'] = options[:headers]['Content-Type'] unless options.dig(:headers, 'Content-Type').nil?

    # Create HTTP::Get or HTTP::Post
    request =  Net::HTTP.const_get(method.capitalize).new(uri.path, request_headers)

    # Add bearer token header
    request['Authorization'] = "Bearer #{@access_token}" if options[:authenticated]

    # Allow caller to modify the request before we send it off:
    yield request if block_given?

    logger.debug "SierraApiClient: #{method} to Sierra api", { uri: uri, body: request.body }

    begin
      # Execute request:
      response = http.request(request)
    rescue => e
      raise SierraApiClientError.new(e), "Failed to #{method} to #{path}: #{e.message}"
    end

    logger.debug "SierraApiClient: Got Sierra api response", { code: response.code, body: response.body }

    parse_response response
  end

  def parse_response (response)
    if response.code == "401"
      # Likely an expired access-token; Wipe it for next run
      # TODO: Implement token refresh
      @access_token = nil
      raise SierraApiClientTokenError.new("Got a 401: #{response.body}")
    end

    SierraApiResponse.new(response)
  end

  def parse_http_options (_options)
    options = {
      authenticated: true
    }.merge _options

    options[:headers] = {
    }.merge(_options[:headers] || {})
      .transform_keys(&:to_s)

    options
  end

  # Authorizes the request.
  def authenticate!
    # NOOP if we've already authenticated
    return nil if ! @access_token.nil?

    logger.debug "SierraApiClient: Authenticating with client_id #{@config[:client_id]}"

    uri = URI.parse("#{@config[:oauth_url]}")
    request = Net::HTTP::Post.new(uri)
    request.basic_auth(@config[:client_id], @config[:client_secret])
    request.set_form_data(
      "grant_type" => "client_credentials"
    )

    req_options = {
      use_ssl: uri.scheme == "https",
      request_timeout: 500
    }

    response = Net::HTTP.start(uri.hostname, uri.port, req_options) do |http|
      http.request(request)
    end

    if response.code == '200'
      @access_token = JSON.parse(response.body)["access_token"]
    else
      nil
    end
  end

  def logger
    @logger ||= NyplLogFormatter.new(STDOUT, level: @config[:log_level])
  end
end


lib/nypl_sierra_api_client.rb
require 'net/http'
require 'net/https'
require 'uri'
require 'nypl_log_formatter'

require_relative 'errors'
require_relative 'sierra_api_response'

class SierraApiClient
  def initialize(config = {})
    config_defaults = {
      env: {
        base_url: 'SIERRA_API_BASE_URL',
        client_id: 'SIERRA_OAUTH_ID',
        client_secret: 'SIERRA_OAUTH_SECRET',
        oauth_url: 'SIERRA_OAUTH_URL'
      },
      static: {
        log_level: 'info'
      }
    }

    @config = config_defaults[:env].map {|k,v| [k, ENV[v]]}.to_h
      .merge config_defaults[:static]
      .merge config

    config_defaults[:env].each do |key, value|
      raise SierraApiClientError.new "Missing config: neither config.#{key} nor ENV.#{value} are set" unless @config[key]
    end

    @retries = 0
  end

  def put (path, body, options = {})
  options = parse_http_options options
    # Default to POSTing JSON unless explicitly stated otherwise
    options[:headers]['Content-Type'] = 'application/json' unless options[:headers]['Content-Type']

    do_request 'put', path, options do |request|
      request.body = body
      request.body = request.body.to_json unless options[:headers]['Content-Type'] != 'application/json'
    end
  end

  def delete (path, options = {})
  options = parse_http_options options

  do_request 'delete', path, options
  end

  def get (path, options = {})
    options = parse_http_options options

    do_request 'get', path, options
  end

  def post (path, body, options = {})
    options = parse_http_options options

    # Default to POSTing JSON unless explicitly stated otherwise
    options[:headers]['Content-Type'] = 'application/json' unless options[:headers]['Content-Type']

    do_request 'post', path, options do |request|
      request.body = body
      request.body = request.body.to_json unless options[:headers]['Content-Type'] != 'application/json'
    end
  end

  private

  def do_request (method, path, options = {})

  raise SierraApiClientError, "Unsupported method: #{method}" unless ['get', 'post', 'put', 'delete'].include? method.downcase

    authenticate! if options[:authenticated]

    @uri = URI.parse("#{@config[:base_url]}#{path}")

    # Build request headers:
    request_headers = {}
    request_headers['Content-Type'] = options[:headers]['Content-Type'] unless options.dig(:headers, 'Content-Type').nil?

    # Create HTTP::Get or HTTP::Post
    request =  Net::HTTP.const_get(method.capitalize).new(@uri, request_headers)

    # Add bearer token header
    request['Authorization'] = "Bearer #{@access_token}" if options[:authenticated]

    # Allow caller to modify the request before we send it off:
    yield request if block_given?

    logger.debug "SierraApiClient: #{method} to Sierra api", { uri: @uri, body: request.body }

    execute request, options
  end

  def execute (request, options)
    http = Net::HTTP.new(@uri.host, @uri.port)
    http.use_ssl = @uri.scheme === 'https'

    begin
      response = http.request(request)
      logger.debug "SierraApiClient: Got Sierra api response", { code: response.code, body: response.body }
    rescue => e
      raise SierraApiClientError.new "Failed to #{request.method} to #{request.path}: #{e.message}"
    end

    handle_response response, request, options
  end

  def handle_response (response, request, options)
    if response.code == "401"
      # Likely an expired access-token; Wipe it for next run
      @access_token = nil
      if @retries < 3
        if options[:authenticated]
          logger.debug "SierraApiClient: Refreshing oauth token for 401", { code: 401, body: response.body, retry: @retries }

          return reauthenticate_and_reattempt request, options
        end
      else
        retries_exceeded = true
      end

      reset_retries
      message = "Got a 401: #{retries_exceeded ? "Maximum retries exceeded, " : ''}#{response.body}"
      raise SierraApiClientTokenError.new(message)
    end

    if response.body == '' && response.code.to_i < 300 && response.code.to_i >= 200
        reattempt_request request, options
    end

    reset_retries if @retries > 0
    SierraApiResponse.new(response)
  end


  def reattempt_request request, options 
    if @retries < 3
      logger.warn "#{request.method} request retry ##{@retries} due to empty response from Sierra API"
      sleep 2 ** (@retries - 1)
      @retries += 1
      execute request, options
    else 
      reset_retries
      raise SierraApiResponseError.new "Sierra API Client: Request failed after 3 empty responses received from Sierra API"
    end 
  end


  def reauthenticate_and_reattempt request, options
    @retries += 1
    sleep 2 ** (@retries - 1)
    authenticate!
    # Reset bearer token header
    request['Authorization'] = "Bearer #{@access_token}"

    execute request, options
  end

  def parse_http_options (_options)
    options = {
      authenticated: true
    }.merge _options

    options[:headers] = {
    }.merge(_options[:headers] || {})
      .transform_keys(&:to_s)

    options
  end

  # Authorizes the request.
  def authenticate!
    # NOOP if we've already authenticated
    return nil if ! @access_token.nil?

    logger.debug "SierraApiClient: Authenticating with client_id #{@config[:client_id]}"

    uri = URI.parse("#{@config[:oauth_url]}")
    request = Net::HTTP::Post.new(uri)
    request.basic_auth(@config[:client_id], @config[:client_secret])
    request.set_form_data(
      "grant_type" => "client_credentials"
    )

    req_options = {
      use_ssl: uri.scheme == "https",
      request_timeout: 500
    }

    response = Net::HTTP.start(uri.hostname, uri.port, req_options) do |http|
      http.request(request)
    end

    if response.code == '200'
      @access_token = JSON.parse(response.body)["access_token"]
    else
      nil
    end
  end

  def logger
    @logger ||= NyplLogFormatter.new(STDOUT, level: @config[:log_level])
  end

  def reset_retries
    @retries = 0
  end
end
